README.md
# Objectives
* To measure the soil moisture of a plant, display the value in Domoticz, in a Tinkerforge Segment Display 4x7 and a Tinkerforge RGB LED indicator.
* To learn how to write generic Python plugin(s) for the Domoticz Home Automation system communicating with [Tinkerforge](http://www.tinkerforge.com) Building Blocks.

![soil-moisture-monitor-p](https://user-images.githubusercontent.com/47274144/58539923-a7e22900-81f8-11e9-8c53-6ed5f8b22fed.png)

## Solution
A Domoticz Python plugin "Soil Moisture Monitor" with a soil moisture monitor device obtaining the moisture value from a Tinkerforge moisture bricklet.
The Tinkerforge moisture bricklet is connected to a Tinkerforge master brick with WiFi extension.
The moisture value is converted to a range 0(dry) - 100(saturated) and displayed in a Tinkerforge segment display 4x7.
In addition, a Tinkerforge rgb led bricklet indicates the state red(dry), yellow(irrigation advice), green(saturated).

![soil-moisture-monitor-c](https://user-images.githubusercontent.com/47274144/58539921-a7499280-81f8-11e9-9586-2403e2a21529.png)

See also file **domoticz-tinkerforge-soilmoisturemonitor-plugin.pdf**.

## Hardware Parts
* Raspberry Pi 3B+ [(Info)](https://www.raspberrypi.org)
* Tinkerforge Master Brick 1.1 [(Info)](https://www.tinkerforge.com/en/doc/Hardware/Bricks/Master_Brick.html#master-brick)
* Tinkerforge WIFI Master Extention 2.0 [(Info)](https://www.tinkerforge.com/en/doc/Hardware/Master_Extensions/WIFI_V2_Extension.html) 
* Tinkerforge Moisture Bricklet 1.1 [(Info)](https://www.tinkerforge.com/en/doc/Hardware/Bricklets/Moisture.html)
* Tinkerforge Segment Display 4X7 Bricklet 1.0 [(Info)](https://www.tinkerforge.com/en/doc/Hardware/Bricklets/Segment_Display_4x7.html)
* Tinkerforge RGB LED Bricklet 1.0 [(Info)](https://www.tinkerforge.com/en/doc/Hardware/Bricklets/RGB_LED.html)

### Notes about the Tinkerforge Moisture Bricklet 1.1
1) In 2018 Tinkerforge stated: The Moisture Bricklet is discontinued and is no longer sold.
An alternate solution could be a Tinkerforge Analog In Bricklet and a 3rd party Soil Moisture sensor.

2) The Tinkerforge Moisture Bricklet Documentation describes the higher the value the more wet (with range 0 - 4095), but ...
field tests show the other way around in a range 2200(wet) - 3400(dry) - this range depends probably on the device,it is used for this solution.

## Software
Versions for developing & using this plugin.
* Raspberry Pi Raspian 4.14
* Domoticz Home Automation System V4.1
* Tinkerforge Python Binding v2.1.22
* Python 3.5.3
* Thonny 3.1.2 (Python IDE)

## Quick Steps
For implementing the Plugin on the Domoticz Server running on the Raspberry Pi.
See also Appendix Python Plugin Code (well documented).

## Tinkerforge Soil Moisture Monitor Prototype
Build the prototype by connecting the Tinkerforge building blocks (see hardware).
Connect the Master Brick to a device running the Brick Deamon and Viewer.
Just in a nutshell the actions taken to setup the Tinkerforge building blocks using the Tinkerforge Brick Viewer.
* Update the devices firmware
* Set the WiFi master extension fixed IP address in client mode
* Obtain the UID's of the Tinkerforge bricklets as required by the Python plugin

After setting up the Tinkerforge building blocks, reset the master brick and check if the master brick canbe reached via WLAN:
ping tf-wifi-ext-ip-address

## Domoticz Web UI's
Open windows Domoticz Setup > Hardware, Domoticz Setup > Log, Domoticz Setup > Devices
This is required to add the new hardware with its device and monitor if the plugin code is running without errors.

## Create folder
```
cd /home/pi/domoticz/plugins/soilmoisturemonitor
```

## Create the plugin
The plugin has a mandatory filename plugin.py located in the newly created plugin folder
For Python development Thonny, running on a Windows 10 device, is used.

Domoticz Python Plugin Source Code: see file **plugin.py**.

## Install the Tinkerforge Python API
There are two options:

### 1) sudo pip install tinkerforge
Advantage: in case of binding updates, only a single folder must be updated.
Check if a subfolder tinkerforge is created in folder /usr/lib/python3/dist-packages.
If not the case, unzip the Tinkerforge Python Binding into the folder /usr/lib/python3/dist-packages.
_Example_
  Create subfolder Tinkerforge holding the Tinkerforge Python Library
```
  cd /home/pi/tinkerforge
```
  Unpack the latest python bindings into folder /home/pi/tinkerforge
```
  Copy /home/pi/tinkerforge to the Python3 dist-packges
  sudo cp -r /home/pi/tinkerforge /usr/lib/python3/dist-packages/
```

In the Python Plugin code amend the import path to enable using the Tinkerforge libraries
```
from os import path
import sys
sys.path
sys.path.append('/usr/lib/python3/dist-packages')
```

### 2) Install the Tinkerforge Python Bindings in a subfolder of the plugin and copy the binding content.
Disadvantage: Every Python plugin using the Tinkerforge bindings must have a subfolder tinkerforge.
In case of binding updates,each of the tinkerforge plugin folders must be updated.
/home/pi/domoticz/plugins/soilmoisturemonitor/tinkerforge

There is no need to amend the path as for option 1.

For either ways, the bindings are used like:
```
import tinkerforge
from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_moisture import BrickletMoisture
from tinkerforge.bricklet_segment_display_4x7 import BrickletSegmentDisplay4x7
from tinkerforge.bricklet_rgb_led import BrickletRGBLED
```
add more depending Tinkerforge brick/bricklet used.

Ensure to update the files in case of newer Tinkerforge Python Bindings.

## Make plugin.py executable
```
cd /home/pi/domoticz/plugins/soilmoisturemonitor
chmod +x plugin.py
```

## Restart Domoticz
Restart Domoticz to find the plugin:
```
sudo systemctl restart domoticz.service
```

**Note**
When making changes to the Python plugin code, ensure to restart Domoticz and refresh any of the Domoticz Web UI's.

## Domoticz Add Hardware Soil Moisure Monitor
**IMPORTANT**
Prior adding, set in the Domoticz Settings the option to allow new hardware.
If this option is not enabled, no new soilmoisture device is created.
Check in the Domoticz log as error message Python script at the line where the new device is used
(i.e. Domoticz.Debug("Device created: "+Devices[1].Name))

In Domoticz Web UI, select tab Setup > Hardware and add the new hardware Soil Moisture Monitor.
The initial check interval is set at 60 seconds. This is a good value for testing, but for finalversion set tohigher value like once per hour (3600 seconds).

## Add Hardware - Check the Domoticz Log
After adding,ensure to check the Domoticz Log (Domoticz Web UI, select tab Setup > Log)
![soil-moisture-monitor-h](https://user-images.githubusercontent.com/47274144/58539922-a7e22900-81f8-11e9-81f4-d54615b13dd6.png)

_Example:_
```
2019-05-29 08:51:45.983 Status: (Soil Moisture Monitor) Started. 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) Debug logging mask set to: PYTHON PLUGIN QUEUE IMAGE DEVICE CONNECTION MESSAGE ALL 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Mode6':'Debug' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Name':'Soil Moisture Monitor' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'DomoticzBuildTime':'2019-05-28 13:05:24' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Version':'1.1.0' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Author':'rwbL' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'HomeFolder':'/home/pi/domoticz/plugins/SoilMoistureMonitor/' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'DomoticzVersion':'4.10841' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Mode5':'60' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Database':'/home/pi/domoticz/domoticz.db' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'UserDataFolder':'/home/pi/domoticz/' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Mode1':'uTP,q2G,zMF' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Port':'4223' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Address':'192.168.1.112' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'HardwareID':'8' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'StartupFolder':'/home/pi/domoticz/' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Mode4':'100' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Language':'en' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'Key':'SoilMoistureMonitor' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) 'DomoticzHash':'5afd0eb96' 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) Device count: 0 
2019-05-29 08:51:46.489 (Soil Moisture Monitor) Creating new Devices 
2019-05-29 08:51:46.490 (Soil Moisture Monitor) Creating device 'Soil Moisture'. 
2019-05-29 08:51:46.491 (Soil Moisture Monitor) Device created: Soil Moisture Monitor - Soil Moisture 
2019-05-29 08:51:46.491 (Soil Moisture Monitor) Creating device 'Status'. 
2019-05-29 08:51:46.492 (Soil Moisture Monitor) Device created: Soil Moisture Monitor - Status 
2019-05-29 08:51:46.492 (Soil Moisture Monitor) Heartbeat set: 60 
2019-05-29 08:51:46.492 (Soil Moisture Monitor) Pushing 'PollIntervalDirective' on to queue 
2019-05-29 08:51:46.492 (Soil Moisture Monitor) Processing 'PollIntervalDirective' message 
2019-05-29 08:51:46.492 (Soil Moisture Monitor) Heartbeat interval set to: 60. 
2019-05-29 08:51:46.486 Status: (Soil Moisture Monitor) Entering work loop. 
2019-05-29 08:51:46.487 Status: (Soil Moisture Monitor) Initialized version 1.1.0, author 'rwbL' 
```
## Domoticz Log Entry Polling with Debug=True
The Soil Moisture Monitor (SMM) runs every 60 seconds (Heartbeat interval) which is shown in the Domoticz log.
```
2019-05-28 19:36:51.236 (Soil Moisture Monitor) Pushing 'onHeartbeatCallback' on to queue 
2019-05-28 19:36:51.240 (Soil Moisture Monitor) Processing 'onHeartbeatCallback' message 
2019-05-28 19:36:51.240 (Soil Moisture Monitor) Calling message handler 'onHeartbeat'. 
2019-05-28 19:36:51.240 (Soil Moisture Monitor) onHeartbeat called. Counter=840 (Heartbeat=60) 
2019-05-28 19:36:51.241 (Soil Moisture Monitor) UIDs:uTP,q2G,zMF 
2019-05-28 19:36:51.253 (Soil Moisture Monitor) IP Connection - OK 
2019-05-28 19:36:51.258 (Soil Moisture Monitor) Tinkerforge value:2918 
2019-05-28 19:36:51.258 (Soil Moisture Monitor) Domoticz value:59 
2019-05-28 19:36:51.258 (Soil Moisture Monitor - Soil Moisture) Updating device from 59:'0' to have values 59:'0'. 
2019-05-28 19:36:51.266 (Soil Moisture Monitor) Segment Display updated 
2019-05-28 19:36:51.266 (Soil Moisture Monitor) RGB LED updated. Brightness=100 
2019-05-28 19:36:51.369 (Soil Moisture Monitor - Status) Updating device from 0:'Polling OK: TF=2918, Dom=59, LED=41' to have values 0:'Polling OK: TF=2918, Dom=59, LED=41'. 
2019-05-28 19:36:51.378 (Soil Moisture Monitor) Polling OK: TF=2918, Dom=59, LED=41 
```

## ToDo
Exception handling for communicating with the Master Brick & WiFi extension.

Consider to replace the Tinkerforge Moisture Bricklet as not produced anymore by an alternate solution using the Tinkerforge Analog In Bricklet and a 3rd party Soil Moisture sensor.

## Version
v1.1.0 (Build 20190529)



CHANGELOG.md
### v1.1.0 (build 20190529)

* NEW: Hardware Bricklet UIDs as Comma Separated String (allows to define additional Bricklets)
* UPD: Improved error handling for IP connection and Bricklet UIDS
* UPD: Plugin definition reworked (Parameter UIDs, LED brightness, Polling)
* UPD: Minor fixes

### v1.0.0 (build 20181205)

* NEW: First version published


plugin.py
# Domoticz Home Automation - SoilMoistureMonitor
# Get the moisture from the Tinkerforge Moisture Bricklet, display in the Segment Display 4x7 Bricklet, state in RGB LED Bricklet
# @author Robert W.B. Linn
# @version 1.1.0 (Build 20190529)
#
# NOTE: after every change run
# sudo chmod +x *.*                      
# sudo systemctl restart domoticz.servicepi
#
# Domoticz Python Plugin Development Documentation:
# https://www.domoticz.com/wiki/Developing_a_Python_plugin

"""
<plugin key="SoilMoistureMonitor" name="Soil Moisture Monitor" author="rwbL" version="1.1.0">
    <description>
        <h2>Soil Moisture Monitor</h2><br/>
        In regular intervals, obtain the moisture value from the Tinkerforge Moisture Bricklet.
        <ul style="list-style-type:square">
            <li>Display the value, 0 (dry) - 100 (wet), in the Tinkerforge Segment Display 4x7 Bricklet</li>
            <li>Indicate the state, red(dry) or yellow(irrigation advice) or green(adequaly wet), in the Tinkerforge RGB LED Bricklet</li>
            <li>Show the value (cp) and state in a Domoticz Soil Moisture device</li>
        </ul>
        <h3>Features</h3>
        <ul style="list-style-type:square">
            <li>Moisture value</li>
            <li>00 - 09 = saturated, 10 - 19 = adequately wet, 20 - 59 = irrigation advice, 60 - 99 = irrigation, 100-200 = Dangerously dry</li>
            <li>LED Indicator: RED=dry <20 or YELLOW=irrigation advice =>20 and <40 or GREEN=adequaly wet > 40</li>
        </ul>
        <h3>Soil Moisture Devices</h3>
        <ul style="list-style-type:square">
            <li>Soil Moisture Value - Soil Moisture device showing the value (cp) and advice</li>
            <li>Soil Moisture Status - Text device to show the latest change or any error condition</li>
        </ul>
        <h3>Configuration</h3>
        Requires the HTTP address and Port of the Master Brick WiFi Extention and the UIDs of the Tinkerforge Bricklets Moisture, Segment Display, RGB LED.<br/>
        The Tinkerforge Bricklet UIDs to be defined as comma separated string of UIDs.
    </description>
    <params>
        <param field="Address" label="Host" width="200px" required="true" default="192.168.N.NNN"/>
        <param field="Port" label="Port" width="75px" required="true" default="4223"/>
        <param field="Mode1" label="UIDs" width="200px" required="true" default="uTP,q2G,zMF"/>
        <param field="Mode4" label="LED Brightness" width="50px" required="true" default="100"/>
        <param field="Mode5" label="Polling (seconds)" width="50px" required="true" default="60"/>
        <param field="Mode6" label="Debug" width="75px">
            <options>
                <option label="True" value="Debug" default="true"/>
                <option label="False" value="Normal"/>
            </options>
        </param>
    </params>
</plugin>
""" 

## Imports
import Domoticz
import urllib
import urllib.request

# Amend the import path to enable using the Tinkerforge libraries
# Alternate (ensure to update in case newer Python API bindings):
# create folder tinkerforge and copy the binding content, i.e.
# /home/pi/domoticz/plugins/soilmoisturemonitor/tinkerforge
from os import path
import sys
sys.path
sys.path.append('/usr/local/lib/python3.5/dist-packages')
                
import tinkerforge
from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_moisture import BrickletMoisture
from tinkerforge.bricklet_segment_display_4x7 import BrickletSegmentDisplay4x7
from tinkerforge.bricklet_rgb_led import BrickletRGBLED

# TF RGB LED digits (according datasheet)
# 0~9,A,b,C,d,E,F
DIGITS = [0x3f,0x06,0x5b,0x4f,
          0x66,0x6d,0x7d,0x07,
          0x7f,0x6f,0x77,0x7c,
          0x39,0x5e,0x79,0x71]

# Set brightness level 0(dark) - 255(brightest)
RGBBRIGHTNESSMIN = 0
RGBBRIGHTNESSMAX = 255

# TF Moisture Bricklet range according test measures
# TF documentation describes the higher the value the more wet (with range 0 - 4095), but ...
# field tests show the other way around in a range 2200(wet) - 3400(dry)
TFMOISTUREDRY = 3400
TFMOISTUREWET = 2200

DOMOTICZMOISTUREDRY = 100
DOMOTICZMOISTUREWET = 0

class BasePlugin:
    
    def __init__(self):
        # The Domoticz heartbeat is set to every 60 seconds. Do not use a higher value as Domoticz message "Error: hardware (N) thread seems to have ended unexpectedly"
        # The Soil Moisture Monitor is read every Parameter.Mode5 seconds. This is determined by using a hearbeatcounter which is triggered by:
        # (self.HeartbeatCounter * self.HeartbeatInterval) % int(Parameter.Mode5) = 0
        self.HeartbeatInterval = 60
        self.HeartbeatCounter = 0
        # Flag to check if connected to the master brick
        self.ipConnected = 0
        return

    def onStart(self):
        Domoticz.Debug("onStart called")
        Domoticz.Debug("Debug Mode:" + Parameters["Mode6"])
        
        if Parameters["Mode6"] == "Debug":
            self.debug = True
            Domoticz.Debugging(1)
            DumpConfigToLog()

        if (len(Devices) == 0):
            # Create new devices for the Soil Moisture Hardware
            Domoticz.Debug("Creating new Devices")
            Domoticz.Device(Name="Soil Moisture", Unit=1, TypeName="Soil Moisture", Used=1).Create()
            Domoticz.Debug("Device created: "+Devices[1].Name)
            Domoticz.Device(Name="Status", Unit=2, TypeName="Text", Used=1).Create()
            Domoticz.Debug("Device created: "+Devices[2].Name)
            
        Domoticz.Debug("Heartbeat set: "+Parameters["Mode5"])
        Domoticz.Heartbeat(self.HeartbeatInterval)

    def onStop(self):
        Domoticz.Debug("Plugin is stopping.")

    def onConnect(self, Connection, Status, Description):
        Domoticz.Debug("onConnect called")
        
    def onMessage(self, Connection, Data):
        Domoticz.Debug("onMessage called")

    def onCommand(self, Unit, Command, Level, Hue):
        Domoticz.Debug("onCommand called for Unit " + str(Unit) + ": Parameter '" + str(Command) + "', Level: " + str(Level))

    def onNotification(self, Name, Subject, Text, Status, Priority, Sound, ImageFile):
        Domoticz.Debug("Notification: " + Name + "," + Subject + "," + Text + "," + Status + "," + str(Priority) + "," + Sound + "," + ImageFile)

    def onDisconnect(self, Connection):
        Domoticz.Debug("onDisconnect called")

    def onHeartbeat(self):
        self.HeartbeatCounter = self.HeartbeatCounter + 1
        Domoticz.Debug("onHeartbeat called. Counter=" + str(self.HeartbeatCounter * self.HeartbeatInterval) + " (Heartbeat=" + Parameters["Mode5"] + ")")
        # Flag to check if connected to the master brick
        self.ipConnected = 0

        # check the heartbeatcounter against the heartbeatinterval
        if (self.HeartbeatCounter * self.HeartbeatInterval) % int(Parameters["Mode5"]) == 0:
            # Get the moisture value
            try:
                # Create IP connection
                ipcon = IPConnection()
            
                # Create device objects using the UID as defined in the parameter Mode1
                # The string contains multiple UIDs separated by comma (,). This enables to define more bricklets.
                brickletUIDParam = Parameters["Mode1"]
                Domoticz.Debug("UIDs:" + brickletUIDParam )
                # Split the parameter string into a list of UIDs
                brickletUIDList = brickletUIDParam.split(',')
                # Check the list length (3 because 3 bricklet UIDs) and create the device objects
                if len(brickletUIDList) == 3:
                    mb = BrickletMoisture(brickletUIDList[0], ipcon)
                    sb = BrickletSegmentDisplay4x7(brickletUIDList[1], ipcon)
                    lb = BrickletRGBLED(brickletUIDList[2], ipcon)

                # Connect to brickd using Host and Port
                try:
                    ipcon.connect(Parameters["Address"], int(Parameters["Port"]))
                    self.ipConnected = 1
                    Domoticz.Debug("IP Connection - OK")
                except:
                    Domoticz.Debug("[ERROR] IP Connection failed")

                # Don't use device before ipcon is connected
                if self.ipConnected == 0:
                    Devices[2].Update( nValue=0, sValue="[ERROR] Can not connect to Master Brick. Check settings, correct and restart Domoticz." )
                    Domoticz.Log(Devices[2].sValue)
                    return

                # Get current moisture value
                moisturetf = mb.get_moisture_value()
                Domoticz.Debug("Tinkerforge value:" + str(moisturetf) )
                moisturedom = converttfvalue(moisturetf)
                Domoticz.Debug("Domoticz value:" + str(moisturedom) )

                # Moisture Device and Text Device
                # Update the value - only nValue is used, but mandatory to add an sValue
                Devices[1].Update( nValue=moisturedom, sValue="0")

                # Tinkerforge Bricklet Updates

                # Segment Display set the value between 0(dry) - 100(wet)
                # Inverse the Domoticz moisure value
                moistureled = DOMOTICZMOISTUREDRY - moisturedom
                l = list(str(moistureled))
                # dry
                if  len(l) == 1:
                    segments = (DIGITS[0], DIGITS[0], DIGITS[0], DIGITS[int(l[0])])

                # irrigation advice
                if  len(l) == 2:
                    segments = (DIGITS[0], DIGITS[0], DIGITS[int(l[0])], DIGITS[int(l[1])])

                # adequate
                if  len(l) == 3:
                    segments = (DIGITS[0], DIGITS[int(l[0])], DIGITS[int(l[1])], DIGITS[int(l[2])])

                # not used
                if  len(l) == 4:
                    segments = (DIGITS[int(l[0])], DIGITS[int(l[1])], DIGITS[int(l[2])], DIGITS[int(l[3])])

                # Write the moisture value to the display with full brightness without colon
                sb.set_segments(segments, 7, False)        
                Domoticz.Debug("Segment Display updated")

                # Set the color of the RGB LED indicator
                # The indicator uses own scheme to keep simple: dry < 20; irrigation advice 20-40; wet > 40
                Domoticz.Debug("RGB LED updated. Brightness=" + Parameters["Mode4"])
                lbbrightness = int(Parameters["Mode4"])
                if lbbrightness < RGBBRIGHTNESSMIN:
                    lbbrightness = RGBBRIGHTNESSMIN
                if lbbrightness > RGBBRIGHTNESSMAX:
                    lbbrightness = RGBBRIGHTNESSMIN

                # Turn the LED on with color RGB depending LED value - 0(dry) - 100(wet)
                # dry = RED
                if moistureled < 20:
                    lb.set_rgb_value(lbbrightness, 0, 0)

                # irrigation advice = YELLOW
                if moistureled >= 20 and moistureled <= 40:
                    lb.set_rgb_value(lbbrightness, lbbrightness, 0)
                
                # wet = GREEN
                if moistureled > 40:
                    lb.set_rgb_value(0, lbbrightness, 0)
               
                # Disconnect
                ipcon.disconnect()

                # Log Message
                Devices[2].Update( nValue=0, sValue="Polling OK: TF=" + str(moisturetf)+", Dom="+str(moisturedom)+", LED="+str(moistureled))
                Domoticz.Log(Devices[2].sValue)

            except:
                # Error
                # Important to close the connection - if not, the plugin can not be disabled
                if self.ipConnected == 1:
                    ipcon.disconnect()
            
                Devices[2].Update( nValue=0, sValue="[ERROR] Check settings, correct and restart Domoticz." )
                Domoticz.Log(Devices[2].sValue)

global _plugin
_plugin = BasePlugin()

def onStart():
    global _plugin
    _plugin.onStart()

def onStop():
    global _plugin
    _plugin.onStop()

def onConnect(Connection, Status, Description):
    global _plugin
    _plugin.onConnect(Connection, Status, Description)

def onMessage(Connection, Data):
    global _plugin
    _plugin.onMessage(Connection, Data)

def onCommand(Unit, Command, Level, Hue):
    global _plugin
    _plugin.onCommand(Unit, Command, Level, Hue)

def onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile):
    global _plugin
    _plugin.onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile)

def onDisconnect(Connection):
    global _plugin
    _plugin.onDisconnect(Connection)

def onHeartbeat():
    global _plugin
    _plugin.onHeartbeat()

# Generic helper functions
def DumpConfigToLog():
    for x in Parameters:
        if Parameters[x] != "":
            Domoticz.Debug( "'" + x + "':'" + str(Parameters[x]) + "'")
    Domoticz.Debug("Device count: " + str(len(Devices)))
    for x in Devices:
        Domoticz.Debug("Device:           " + str(x) + " - " + str(Devices[x]))
        Domoticz.Debug("Device ID:       '" + str(Devices[x].ID) + "'")
        Domoticz.Debug("Device Name:     '" + Devices[x].Name + "'")
        Domoticz.Debug("Device nValue:    " + str(Devices[x].nValue))
        Domoticz.Debug("Device sValue:   '" + Devices[x].sValue + "'")
        Domoticz.Debug("Device LastLevel: " + str(Devices[x].LastLevel))
    return

# Convert TF value to Domoticz value between 0(wet) - 100(dry)
# TF: # 2200(wet) - 3400(dry)
# Domoticz: 0(wet) - 100(dry) (> 100 is not used)
# Domoticz: 00 - 09 = saturated, 10 - 19 = adequately wet, 20 - 59 = irrigation advice, 60 - 99 = irrigation, 100-200 = Dangerously dry
def converttfvalue(tfvalue):
    v = 0
    # Avoid dividing by 0
    if tfvalue > TFMOISTUREWET:
        r = TFMOISTUREDRY - TFMOISTUREWET
        # Set the domoticz value using the tf percentage factor of the tf value in the range
        f = (tfvalue - TFMOISTUREWET) / r
        v = DOMOTICZMOISTUREDRY * f
    return int(v)
